### PR TITLE
First draft TabularMSA repr

### LINKED ISSUE #1172: `TabularMSA`: __repr__
Not sure what this looks like yet.

### PR DESCRIPTION
**Pull request by anderspitman**
Closes #1172

### CODE REVIEW

**Comment by anderspitman** [comment]
Most of the real work is still left to be done, which includes:

- [ ]  Major refactoring to remove duplication betwen `_SequenceReprBuilder` and `_TabularMSAReprBuilder`
- [ ] Move metadata and positional metadata repr code into a separate place
- [ ] Write tests for `TabularMSAReprBuilder`

**Comment by anderspitman** [comment]
@jairideout ready for review.

**Comment by jairideout** [comment]
Thanks @anderspitman! Can you also add `__str__`, which just calls `__repr__`? #1171

File: skbio/alignment/_tabular_msa.py [hunk 1]
```diff
@@ -22,6 +22,8 @@
 from skbio.util._decorator import experimental, classonlymethod, overrides
 from skbio.util._misc import resolve_key
 
+from skbio.alignment._tabular_msa_repr import _TabularMSAReprBuilder
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: recommend renaming `skbio.alignment._tabular_msa_repr` to `skbio.alignment._repr` to match `skbio.sequence._repr`

File: skbio/alignment/_tabular_msa.py [hunk 2]
```diff
@@ -259,6 +261,21 @@ def __init__(self, sequences, metadata=None, positional_metadata=None,
...
+    def __repr__(self):
+        pep8_line_length_limit = 79
+        length_taken_by_docstring_indent = 8
+        width = pep8_line_length_limit - length_taken_by_docstring_indent
+        return _TabularMSAReprBuilder(
+            seq=self,
+            width=width,
+            indent=4).build()
+
+    @experimental(as_of='0.4.0-dev')
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: remove, not necessary for private functionality

File: skbio/alignment/_tabular_msa.py [hunk 3]
```diff
@@ -259,6 +261,21 @@ def __init__(self, sequences, metadata=None, positional_metadata=None,
...
+    def _repr_stats(self):
+        return [("Sequence Count", str(self.shape.sequence)),
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: lowercase these for consistency with `Sequence` stats:

"sequence count"
"position count"

File: skbio/alignment/_tabular_msa_repr.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class _TabularMSAReprBuilder(_MetadataReprBuilder):
+
+    def _process_header(self):
+        cls_name = self._obj.__class__.__name__
+        if self._obj.dtype is not None:
+            dtype_class_name = self._obj.dtype.__name__
+        else:
+            dtype_class_name = 'No dtype'
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: How about just displaying `TabularMSA` without the dtype info if there isn't a dtype? `TabularMSA<No dtype>` looks a little odd to me.

File: skbio/alignment/_tabular_msa_repr.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        self._lines.add_line(cls_name + ' <' + dtype_class_name + '>')
+        self._lines.add_separator()
+
+    def _process_data(self):
+        num_lines = self._obj.shape.sequence
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: The names here are kind of confusing. Can you rename `num_lines` to `num_seqs` or `num_sequences`? `_format_lines` probably makes sense to be named something like `_format_sequences(self, sequence_indices)`.

File: skbio/alignment/_tabular_msa_repr.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+        # display entire sequence if we can, else display the first two and
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: *display all sequences if we can, else display the first two and last two sequences separated by ellipsis*

File: skbio/alignment/_tabular_msa_repr.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            self._lines.add_lines(self._format_lines(range(2)))
+            self._lines.add_line('...')
+            self._lines.add_lines(self._format_lines(range(num_lines - 2,
+                                                     num_lines)))
+
+    def _format_lines(self, line_indices):
+        num_chars_at_end = 4
+        lines = []
+        for line_index in line_indices:
+            seq_str = str(self._obj._seqs[line_index])
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: Use `TabularMSA._get_sequence`, indexing the Series this way can give the wrong result depending on index type

File: skbio/alignment/_tabular_msa_repr.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        if num_lines <= 5:
+            self._lines.add_lines(self._format_lines(range(num_lines)))
+        else:
+            self._lines.add_lines(self._format_lines(range(2)))
+            self._lines.add_line('...')
+            self._lines.add_lines(self._format_lines(range(num_lines - 2,
+                                                     num_lines)))
+
+    def _format_lines(self, line_indices):
+        num_chars_at_end = 4
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: Sorry this wasn't clear in the repr spec, but the head and tail of the sequence should be the same number of characters roughly (it looks like the current implementation always displays the last 4 characters regardless of sequence length).

File: skbio/alignment/tests/test_tabular_msa.py [hunk 9]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+        # basic sanity checks -- more extensive testing of formatting and
+        # special cases is performed in TabularMSAReprDoctests below. here we
+        # only test that pieces of the repr are present. these tests also
+        # exercise coverage for py2/3 since the doctests in
+        # TabularMSAReprDoctests only currently run in py3.
+
+        # minimal
+        obs = repr(TabularMSA([DNA('')]))
+        self.assertEqual(obs.count('\n'), 6)
+        self.assertTrue(obs.startswith('TabularMSA'))
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: Check for dtype too?

File: skbio/alignment/tests/test_tabular_msa.py [hunk 10]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
 
+class TestRepr(unittest.TestCase):
+    def test_repr(self):
+        # basic sanity checks -- more extensive testing of formatting and
+        # special cases is performed in TabularMSAReprDoctests below. here we
+        # only test that pieces of the repr are present. these tests also
+        # exercise coverage for py2/3 since the doctests in
+        # TabularMSAReprDoctests only currently run in py3.
+
+        # minimal
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: Can you add a test for an MSA with zero sequences?

File: skbio/alignment/tests/test_tabular_msa.py [hunk 11]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+# will not run the unit tests otherwise
+#
+# these doctests exercise the correct formatting of TabularMSA's repr in a
+# variety of situations. they are more extensive than the unit tests above
+# (TestRepr.test_repr) but are only currently run in py3. thus, they cannot
+# be relied upon for coverage (the unit tests take care of this)
+class TabularMSAReprDoctests(object):
+    r"""
+    >>> import pandas as pd
+    >>> from skbio import Sequence
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: unnecessary import

File: skbio/alignment/tests/test_tabular_msa.py [hunk 12]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+# NOTE: this must be a *separate* class for doctests only (no unit tests). nose
+# will not run the unit tests otherwise
+#
+# these doctests exercise the correct formatting of TabularMSA's repr in a
+# variety of situations. they are more extensive than the unit tests above
+# (TestRepr.test_repr) but are only currently run in py3. thus, they cannot
+# be relied upon for coverage (the unit tests take care of this)
+class TabularMSAReprDoctests(object):
+    r"""
+    >>> import pandas as pd
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: unnecessary import

File: skbio/alignment/_tabular_msa_repr.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+class _TabularMSAReprBuilder(_MetadataReprBuilder):
+
+    def _process_header(self):
+        cls_name = self._obj.__class__.__name__
+        if self._obj.dtype is not None:
+            dtype_class_name = self._obj.dtype.__name__
+        else:
+            dtype_class_name = 'No dtype'
+        self._lines.add_line(cls_name + ' <' + dtype_class_name + '>')
```
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: Can you remove the space between `TabularMSA` and `<`?

```
TabularMSA<DNA>
```

File: skbio/alignment/tests/test_tabular_msa.py [hunk 14]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+
+    >>> TabularMSA([])
+    TabularMSA <No dtype>
+    ---------------------
+    Stats:
+        Sequence Count: 0
+        Position Count: 0
+    ---------------------
+
+    MSA with single empty sequence:
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: add newline here

File: skbio/alignment/tests/test_tabular_msa.py [hunk 15]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+    >>> TabularMSA([DNA('')])
+    TabularMSA <DNA>
+    ---------------------
+    Stats:
+        Sequence Count: 1
+        Position Count: 0
+    ---------------------
+    <BLANKLINE>
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: This has an extra blank line, can you update so it matches the empty `TabularMSA([])` case above?

File: skbio/alignment/tests/test_tabular_msa.py [hunk 16]
```diff
@@ -2211,5 +2211,181 @@ def test_negative_int(self):
...
+    Stats:
+        Sequence Count: 6
+        Position Count: 71
+    -----------------------------------------------------------------------
+    AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+    AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+    ...
+    AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+    AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+    """
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: Can you add a test for an MSA with `metadata` and `positional_metadata`?

File: skbio/util/_metadata_repr.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+from __future__ import absolute_import, division, print_function
+from future.utils import with_metaclass
+
+import six
+import itertools
+import numbers
+import textwrap
+
+from abc import ABCMeta, abstractmethod
+from skbio.sequence._base import ElasticLines
```
Comment (jairideout) [inline skbio/util/_metadata_repr.py]: Can you move `ElasticLines` into `skbio._base` (as well as the unit tests)? It's being used now within `sequence` and `alignment`.

File: skbio/util/_metadata_repr.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+    Parameters
+    ----------
+    obj : Type varies depending on subclass
+        Object to build repr for.
+    width : int
+        Maximum width of the repr.
+    indent : int
+        Number of spaces to use for indented lines.
+    chunk_size: int
```
Comment (jairideout) [inline skbio/util/_metadata_repr.py]: remove, this parameter doesn't exist here

File: skbio/util/_metadata_repr.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+        Number of characters in each chunk of a sequence.
+
+    """
+    def __init__(self, seq, width, indent):
```
Comment (jairideout) [inline skbio/util/_metadata_repr.py]: rename `seq` to `obj`

**Comment by jairideout** [comment]
@anderspitman looks great! Mostly minor comments, let me know when you're ready for another pass

File: skbio/alignment/_tabular_msa_repr.py (hunk 13 above)
Comment (anderspitman) [inline skbio/alignment/_tabular_msa_repr.py]: Will do, although it was intentional. I feel like it looks better. If you haven't compared side by side you might consider taking a look.

**Comment by anderspitman** [comment]
@jairideout I'll try to have this back in the next hour or two.

File: skbio/alignment/_tabular_msa.py (hunk 2 above)
Comment (anderspitman) [inline skbio/alignment/_tabular_msa.py]: It looks like the other dunders are decorated?

Comment (anderspitman) [inline skbio/alignment/_tabular_msa.py]: nvm, that's not a dunder

File: skbio/alignment/_tabular_msa_repr.py (hunk 7 above)
Comment (anderspitman) [inline skbio/alignment/_tabular_msa_repr.py]: Nice, I was wishing for something like and didn't realize we already had it.

File: skbio/alignment/_tabular_msa_repr.py (hunk 13 above)
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: We were thinking of using parameterized type notation, e.g. similar to C++ and Java:

```java
ArrayList<Integer> numbers = new ArrayList<Integer>();
```

I compared both styles and like it without the space because it matches the style above and seems to make a stronger statement about it being a `TabularMSA` *of* `DNA` sequences.

@ebolyen what's your preference?

File: skbio/util/_metadata_repr.py (hunk 17 above)
Comment (anderspitman) [inline skbio/util/_metadata_repr.py]: `skbio.sequence._base.py` is empty after this. I removed it and the test file.

File: skbio/alignment/_tabular_msa_repr.py (hunk 13 above)
Comment (ebolyen) [inline skbio/alignment/_tabular_msa_repr.py]: :+1: for no space, but looking into this further, check out this PEP:
https://www.python.org/dev/peps/pep-0484/

It suggests using the mypy strategy of using square braces. This might be more pythonic:
`TabularMSA[DNA]`. (It's actually a pretty cool way of describing higher-kinded types, use `__getitem__` as the constructor instead of a function, might warrant some thought for qiime 2)

Comment (anderspitman) [inline skbio/alignment/_tabular_msa_repr.py]: I like the comparison to C++/Java. I say remove it.

File: skbio/util/_metadata_repr.py (hunk 17 above)
Comment (jairideout) [inline skbio/util/_metadata_repr.py]: :+1:

File: skbio/alignment/_tabular_msa_repr.py (hunk 13 above)
Comment (jairideout) [inline skbio/alignment/_tabular_msa_repr.py]: :+1: for using square brackets to be more pythonic: `TabularMSA[DNA]`

File: skbio/alignment/_repr.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _process_header(self):
+        cls_name = self._obj.__class__.__name__
+        if self._obj.dtype is not None:
+            dtype_class = '<' + self._obj.dtype.__name__ + '>'
+        else:
+            dtype_class = ''
+        self._lines.add_line(cls_name + dtype_class)
+        self._lines.add_separator()
+        self._ellipse_insert = ' ... '
```
Comment (jairideout) [inline skbio/alignment/_repr.py]: It's a little odd to set `_ellipse_insert` here, I think this makes more sense in the constructor

**Comment by jairideout** [comment]
@anderspitman I took another pass through, the last changes are to display `TabularMSA[DNA]` and I had one more inline comment about `_ellipse_insert`. Once those are addressed this is ready for merge!

File: skbio/alignment/_repr.py (hunk 20 above)
Comment (anderspitman) [inline skbio/alignment/_repr.py]: That was a mistake. Thought that was the constructor for some reason...

**Comment by jairideout** [comment]
Closing in favor of #1211 which contains the commits here.
